torch/nn/modules/transformer.py

```diff
@@ -227,7 +246,11 @@ def forward(self, tgt: Tensor, memory: Tensor, tgt_mask: Optional[Tensor] = None
         Shape:
             see the docs in Transformer class.
         """
-        output = tgt
+        if self.batch_first:
```

Comment:
Do we want to have the transpose on the transformer level or the MHA level?

Reply:
Locally on my end, I followed implementation here [https://nlp.seas.harvard.edu/2018/04/03/attention.html](https://nlp.seas.harvard.edu/2018/04/03/attention.html) and changed the batch first in the MHA level so that the input to MHA is `(batch, n_head, seq_len, d_model//n_head)`. But I thought it would be simpler if just change it in the wrapper level.

Reply:
I think it's fairly common to use MHA directly without the Transformer layers, since they're not flexible enough yet for certain use cases. Given that, I think it'd be good to support `batch_first` at the MHA level.